Sources/SourceKit/SourceKitServer.swift

```diff
@@ -418,13 +458,17 @@ extension SourceKitServer {
   // MARK: - Text synchronization
 
   func openDocument(_ note: Notification<DidOpenTextDocumentNotification>, workspace: Workspace) {
-    workspace.documentManager.open(note.params)
+    openDocument(note.params, workspace: workspace)
+  }
+
+  private func openDocument(_ note: DidOpenTextDocumentNotification, workspace: Workspace) {
+    workspace.documentManager.open(note)
 
-    let textDocument = note.params.textDocument
+    let textDocument = note.textDocument
     workspace.buildSettings.registerForChangeNotifications(for: textDocument.uri)
```

Comment:
I guess, I should close the document the same way as I open it, by sending a `DidCloseTextDocumentNotification`. That should take care of unregistering it as well.

Reply:
Actually on second thought, we should probably just not re-register for change notifications. As far as `sourcekit-lsp` is concerned, the files are still open, there has just been a hick-up with the underlying functionality. As far as the build system is concerned, we can still receive updated build settings and they will be used when `sourcekitd` restarts (see the comment about not resetting the `buildSettingsByFile`). 

What do you think?

Reply:
Yeah, that makes sense, but we'll need to make sure that each `ToolchainLanguageServer` correctly receives the new build settings or caches it (including clangd)

Reply:
Since the restarted clang language server is being initialized with the same build settings as before (https://github.com/apple/sourcekit-lsp/blob/master/Sources/SourceKit/clangd/ClangLanguageServer.swift#L211) and `SwiftLanguageServer` sends the build settings on document open (https://github.com/apple/sourcekit-lsp/blob/master/Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift#L254), we should be safe here. Might be worth to add a test case for though.

Reply:
Ah right as long as long as the `ClangLanguageServerShim` receives the new open document request it should be OK. Yeah a test for this seems good

Reply:
Added a test case that verifies that the build settings are recovered after a crash.